src/coreclr/jit/flowgraph.cpp

```diff
@@ -4184,3 +4191,321 @@ void Compiler::fgLclFldAssign(unsigned lclNum)
         lvaSetVarDoNotEnregister(lclNum DEBUGARG(DoNotEnregisterReason::LocalField));
     }
 }
+
+//------------------------------------------------------------------------------
+// fgExpandThreadLocalAccess : Expand the CORINFO_HELP_GETSHARED_NONGCTHREADSTATIC_BASE_NOCTOR_OPTIMIZED
+//                             that access fields marked with [ThreadLocal].
+//
+// Returns:
+//    PhaseStatus indicating what, if anything, was changed.
+//
+// Notes:
+//    A cache is stored in thread local storage (TLS) of coreclr. It maps the typeIndex (embedded in
+//    the code at the JIT time) to the base of static blocks. This method generates code to
+//    extract the TLS, get the entry at which the cache is stored. Then it checks if the typeIndex of
+//    enclosing type of current field is present in the cache and if yes, extract out that can be directly
+//    accessed at the uses.
+//    If the entry is not present, the helper is called, which would make an entry of current static block
+//    in the cache.
+//
+PhaseStatus Compiler::fgExpandThreadLocalAccess()
+{
+    PhaseStatus result = PhaseStatus::MODIFIED_NOTHING;
+
+    if (!doesMethodHaveTlsFieldAccess())
+    {
+        // The method doesn't have any TLS field
+        return result;
+    }
+
+    CORINFO_THREAD_STATIC_BLOCKS_INFO threadStaticBlocksInfo;
+    info.compCompHnd->getThreadLocalStaticBlocksInfo(&threadStaticBlocksInfo);
+    assert(threadStaticBlocksInfo.tlsIndex.accessType == IAT_VALUE);
+
+    for (BasicBlock* block = fgFirstBB; block != nullptr; block = block->bbNext)
+    {
+    SCAN_BLOCK_AGAIN:
+        for (Statement* const stmt : block->Statements())
+        {
+            if ((stmt->GetRootNode()->gtFlags & GTF_CALL) == 0)
+            {
+                // TP: Stmt has no calls - bail out
+                continue;
+            }
+
+            for (GenTree* const tree : stmt->TreeList())
+            {
+                // We only need calls with IsExpRuntimeLookup() flag
+                if (!tree->IsCall() || !tree->AsCall()->IsHelperCall())
```

Comment:
seems like `IsHelperCall` is enough? + one more `IsHelperCall` check below

Reply:
I should remove check from here, but to know if `IsHelperCall()` is true or not, we need to find out if it is a `GT_CALL` or not and hence `IsCall()` check.

Reply:
> but to know if IsHelperCall() is true or not, we need to find out if it is a GT_CALL or not and hence IsCall() check.

There is a GenTree version that checks both: https://github.com/dotnet/runtime/blob/2c8cb129ed084e374f06eaa29f21e9ab0b9e8066/src/coreclr/jit/gentree.h#L9822

Reply:
ah, thanks for that.